### PR TITLE
More (much more) I/O content

### PR DESCRIPTION
**Pull request by eguiraud**
- add section about `ClassDef` to "I/O of custom classes"
- revise "I/O Concepts" section of the manual, make it publicly visible

### CODE REVIEW

File: manual/io/index.md
Comment (pcanal) [inline manual/io/index.md:21]: What is the process to maintain the commit information that is embedded here? (I.e. if TDirectory.h changes, how do we remember to update this commit number)?

File: manual/io/index.md
Comment (pcanal) [inline manual/io/index.md:157]:
```suggestion
To accommodate doxygen-style documentation, this annotation can also be written as `///<!`. 
```

File: manual/io/index.md
Comment (eguiraud) [reply manual/io/index.md:21]: the CI at `root-project/web` complains next time anyone tries to modified the website

**Review by pcanal — APPROVED** [review]
Very nice.  Thanks.   The doxygen-friendly annotation needs to be reviewed.

File: manual/io/index.md
Comment (eguiraud) [reply manual/io/index.md]: Thanks, does this mean that `///->` should be `///<->`?

Comment (pcanal) [reply manual/io/index.md]: Looks very weird but I guess so (@couet @Axel-Naumann)?

**Review by ikabadzhov — APPROVED** [review]

File: manual/io_custom_classes/index.md
Comment (vepadulano) [inline manual/io_custom_classes/index.md:209]:
```suggestion
The `ClassDef` macro can be inserted in a class definition to add some reflection capabilities to it. It also attaches a "version number" to the class that can be used for [schema evolution]({{ '/manual/io/#dealing-with-changes-in-class-layouts-schema-evolution' | relative_url }}).
```

File: manual/io_custom_classes/index.md
Comment (vepadulano) [inline manual/io_custom_classes/index.md:236]: Looks arbitrary. I guess there is an explanation and we don't want to expose it? A sentence above says `A version number equal to 0 tells ROOT to not store the class in a root file`, but no explanation for version 1 or 2

File: manual/io/index.md
Comment (vepadulano) [inline manual/io/index.md:57]: `TDirectory`-s : the dash looks out of place.

File: manual/io/index.md
Comment (vepadulano) [inline manual/io/index.md:66]:
```suggestion
  // OR (less preferred):
```

File: manual/io/index.md
Comment (vepadulano) [inline manual/io/index.md:67]:
```suggestion
  hist->Write(); // the key will be named as the result of `hist->GetName()` inside `gDirectory`
```

File: manual/io/index.md
Comment (vepadulano) [inline manual/io/index.md:77]: It is clear that `Delete` does not delete the object on disk. It is not clear how to delete the object on disk. Maybe it's not the point of the sentence though

File: manual/io/index.md
Comment (vepadulano) [inline manual/io/index.md:103]:
```suggestion
for (auto key: ROOT::RangeStaticCast<TKey*>(*f.GetListOfKeys())) {
```

File: manual/io/index.md
Comment (vepadulano) [inline manual/io/index.md:151]: First mention of `rootcling` with no links or a few words to explain. Might be that we expect readers to know/understand what this is already

File: manual/io/index.md
Comment (vepadulano) [inline manual/io/index.md:183]:
```suggestion
The pointer data member `fH` is marked as never null: ROOT will be able to perform additional optimizations.
```

File: manual/io/index.md
Comment (vepadulano) [inline manual/io/index.md:209]:
```suggestion
> Pointers to simple types (e.g. `float*`, `int*`) are assumed to be variable-size arrays. 
```

File: manual/io/index.md
Comment (vepadulano) [inline manual/io/index.md:279]: Where should `IgnoreObjectStreamer` be invoked? Some place in the class definition? Or in the user application that is writing objects of that class to a file?

**Review by vepadulano — COMMENTED** [review]
Thanks a lot for this! It is a very interesting read, with details that I didn't know. A few comments from my side to make it more clear for readers

File: manual/io/index.md
Comment (jalopezg-r00t) [inline manual/io/index.md:109]:
```suggestion
The output of such an `iterate.C` ROOT macro could be:
```

File: manual/io/index.md
Comment (jalopezg-r00t) [inline manual/io/index.md:138]:
```suggestion
See also → [The `ClassDef` Macro]({{ '/manual/io_custom_classes/#the-classdef-macro' | relative_url }})).
```

File: manual/io/index.md
Comment (jalopezg-r00t) [inline manual/io/index.md:241]:
```suggestion
If `nbits` is present, the member is saved with `nbits`-bits precision. For more details, see [this tutorial](https://root.cern/doc/master/double32_8C.html).
```

File: manual/io/index.md
Comment (jalopezg-r00t) [inline manual/io/index.md:279]: I think this should be:
```suggestion
Classes that do inherit from `TObject` can exclude TObject's data members from their I/O invoking `MyClass->Class()->IgnoreTObjectStreamer()`.
```

File: manual/io/index.md
Comment (jalopezg-r00t) [inline manual/io/index.md:393]: `d` name collides with `d` member declared in line 395.
```suggestion
   char chr; // new data member added
```

File: manual/io/index.md
Comment (jalopezg-r00t) [inline manual/io/index.md:449]:
```suggestion
The schema evolution C++ API consists of the following classes:
```

File: manual/io/index.md
Comment (jalopezg-r00t) [inline manual/io/index.md:459]: I guess `R__b` is the `TBuffer` argument, but do we refer to `R__v` anywhere else?

**Review by jalopezg-r00t — APPROVED** [review]
It's much better like this!  I have attached some comments; feel free to apply before merging.  :slightly_smiling_face:

File: manual/io/index.md
Comment (eguiraud) [reply manual/io/index.md]: @couet confirmed it's `///<->`, fixed

File: manual/io_custom_classes/index.md
Comment (eguiraud) [reply manual/io_custom_classes/index.md:236]: @Axel-Naumann mentioned when writing the document that 3 is actually a _better_ initial version number than others. We need his help here.

File: manual/io/index.md
Comment (eguiraud) [reply manual/io/index.md:57]: it's because `TDirectories` is weird and `TDirectory`s is wrong.

I'll go with `TDirectories` as the lesser evil.

File: manual/io/index.md
Comment (eguiraud) [reply manual/io/index.md:66]: I don't think so, why?

File: manual/io/index.md
Comment (eguiraud) [reply manual/io/index.md:103]: we actually discussed this and decided to go with explicit types (e.g. here to make it clear that `RangeStaticCast<TKey*>` returns an iterator over `TKey *`, and not `TKey&` for instance

File: manual/io/index.md
Comment (eguiraud) [reply manual/io/index.md:77]: Javi's understanding is correct, and it should _not_ be clear that `Delete` does not delete the object on disk, because it does :smile: it just does not free the corresponding disk space. Should we clarify somewhere, @vepadulano ?

File: manual/io/index.md
Comment (eguiraud) [reply manual/io/index.md:151]: at the beginning of this document we say it "assumes that you have read the introduction on → [ROOT files]({{ '/manual/root_files/' | relative_url }}), → [I/O of custom classes]({{ '/manual/io_custom_classes/' | relative_url }}), and → [Trees]({{ '/manual/trees/' | relative_url }})".

In particular "I/O of custom classes" talks about rootcling

Comment (eguiraud) [reply manual/io/index.md:151]: I added a link to the section @jalopezg-r00t suggested, I don't think we have anything better

File: manual/io/index.md
Comment (eguiraud) [reply manual/io/index.md:279]: @pcanal @Axel-Naumann is the following correct?
```suggestion
Classes that do inherit from `TObject` can exclude TObject's data members from their I/O invoking `MyClass->Class::IgnoreObjectStreamer()` before any object of type `MyClass` is written to a ROOT file.
```

Comment (eguiraud) [reply manual/io/index.md:279]: You are right, and the instance should not be called like a type:

```suggestion
Classes that do inherit from `TObject` can exclude TObject's data members from their I/O invoking `myObject->Class()->IgnoreObjectStreamer()`.
```

File: manual/io/index.md
Comment (eguiraud) [reply manual/io/index.md:459]: we don't say what `R__b` is, nor `R__v` (the version number). Rather than just _saying_ it I'd like to point to some reference or some code that shows what signature you can expect a streamer function to have. @pcanal help? :smile:

File: manual/io/index.md
Comment (vepadulano) [reply manual/io/index.md:66]: For style reasons I believe, but it's not necessary

File: manual/io/index.md
Comment (vepadulano) [reply manual/io/index.md:77]: I think I got distracted by some issue that I remember in the past where `Delete`ing an object from a TFile would not really delete it. As in, running a program that just deletes an object (and does not do anything else with the file, thus it does not fill the space that was previously occupied by that object) and then running another program that lists the content of the same file, would show the previously deleted object. Maybe I just misrember, I haven't tried to reproduce this myself again

Comment (eguiraud) [reply manual/io/index.md:77]: please open a github issue if that's the case :smiley:

Comment (vepadulano) [reply manual/io/index.md:77]: Now I remember what was my source of confusion:
```cpp
#include <TFile.h>
#include <TH1.h>

void write(){
    TFile f{"myfile.root", "recreate"};
    TH1F h1{"h1","h1",10,0,10};
    TH1F h2{"h2","h2",10,0,10};
    f.WriteObject(&h1,"");
    f.WriteObject(&h2,"");
}

void read_and_delete(){
    TFile f{"myfile.root","update"};
    f.Delete("h1"); // change to  f.Delete("h1;1") to actually remove the object from the TFile
}

void read_and_list(){
    TFile f{"myfile.root","read"};
    f.ls();
}

int main(){
    write();
    read_and_delete();
    read_and_list();
}
```

In the `read_and_delete` function above, the "h1" histogram is only actually deleted from the TFile if the full namecycle is indicated `h1;1`. If only the name of the histogram is indicated, the `Delete` function deletes it from memory (i.e. for the lifetime of the current open file), but then closing and reopening the file again shows both histograms still there.

Comment (vepadulano) [reply manual/io/index.md:77]: I'm thinking either clearer documentation in `TDirectory::Delete` or a concrete example here in the manual, or both things

Comment (eguiraud) [reply manual/io/index.md:77]: It's documented at https://root.cern/doc/master/classTDirectoryFile.html#a24747f020c3276e8a7d2e6c91e5e3f76

Comment (eguiraud) [reply manual/io/index.md:77]: (and also https://root.cern/doc/master/classTDirectory.html#a09e51150c19fd5a0aa9278a4158202ca )

Comment (eguiraud) [reply manual/io/index.md:77]: (i'm not saying that the behavior is intuitive, but I don't think there is more to be done here in the manual besides linking to those docs)

Comment (vepadulano) [reply manual/io/index.md:77]: Yes, fair enough :+1:

File: manual/io_custom_classes/index.md
Comment (pcanal) [reply manual/io_custom_classes/index.md:236]: Here is the full reason for this seemingly arbitrary number:

A class layout can be either versioned (using a `ClassDef` or `#pragma link C++ options=version(3) class MyClass` or `<class name="pat::Muon"  ClassVersion="12">` or the macro `RootClassVersion`) or unversioned (in which case the class layout checksum is stored and used instead).

However when an unversioned class layout is loaded (either via the dictionary or via `TStreamerInfo`s for a file), they are given a version number starting at 1.

So if a class at any point in its development life unversioned and later becomes versioned and the user want to be able to load "older" files that used the unversioned class layout, the user need to leave some 'space' in the `TStreamerInfo` array (indexed on the version number) for the unversioned class layout.

Consequently the "real" advice is to start the versioning at "number of distinct unversioned class layout" that you might need to load at once (in a process that has this class now versioned).  

To "keep it simple", we came up with a compromise (the value 3) that was "safe enough" (Can read at the same time old files from 2 distinct eras of the life of that class) while not wasting space (if no files with unversioned class layout are read those first few array slot are "wasted" memory usage).

File: manual/io/index.md
Comment (pcanal) [reply manual/io/index.md]: > before any object of type `MyClass` is written to a ROOT file.

That is correct.

Comment (eguiraud) [reply manual/io/index.md]: Added!

File: manual/io/index.md
Comment (pcanal) [reply manual/io/index.md:459]: For R__b, this is the argument to the `Streamer` function, usually declared by `ClassDef`:
```
virtual_keyword void Streamer(TBuffer &R__b) ...
```
R__v is a local variable usually assigned via:
```
   if (b.IsReading()) {
      UInt_t R__s, R__c;
      Version_t R__v = b.ReadVersion(&R__s, &R__c);
      ...... actual custom streamer (often delegating to the default streamer)
      b.CheckByteCount(R__s, R__c, TTree::IsA());
   }
```

File: manual/io_custom_classes/index.md
Comment (jalopezg-r00t) [reply manual/io_custom_classes/index.md:236]:
```suggestion
   // Note that the initial version number be greater than the number of previously, unnumbered (i.e. lacking a
   // explicit `ClassDef`) versions of the class. If unsure, `3` is typically a good compromise.
```

File: manual/io/index.md
Comment (jalopezg-r00t) [inline manual/io/index.md:460]: @pcanal That should be it, no?
```suggestion
   // `buf` below is the TBuffer& argument to the `Streamer` function
   UInt_t R__s, R__c;
   Version_t R__v = buf.ReadVersion(&R__s, &R__c);

   Event::Class()->ReadBuffer(buf, this);
   if (R__v > 1) buf >> fNew;
   else fNew = 0;        // set to some default value

   buf.CheckByteCount(R__s, R__c, Event::IsA());
```

**Review by jalopezg-r00t — COMMENTED** [review]
That should be all that is required to merge the PR.

@eguiraud If @pcanal acknowledges, I think we can merge this.  :-)

File: manual/io/index.md
Comment (pcanal) [reply manual/io/index.md:460]: humm .. looks odd.

Comment (pcanal) [reply manual/io/index.md:460]: Right, the odd part is the mix of 'hand coded streamer and automatic streamer.  I would simply replace
```
Event::Class()->ReadBuffer(buf, this);
```
with
```
...
```

Also note that `TClass::ReadBuffer(TBuffer&, void*)` will read the class version and check the byte count (i.e. fail in this case, since the buffer already advanced passed the version).

Comment (jalopezg-r00t) [reply manual/io/index.md:460]: > Right, the odd part is the mix of 'hand coded streamer and automatic streamer. I would simply replace
> 
> ```
> Event::Class()->ReadBuffer(buf, this);
> ```
> 
> with
> 
> ```
> ...
> ```

I have edited the suggestion and added a more explicit comment.  If there are no other suggestions, we could merge this PR.

Comment (pcanal) [reply manual/io/index.md:460]: Go ahead.  It looks good.

**Comment by eguiraud** [comment]
Rebased to fix conflicts, will merge if green. Thank you @JavierCVilla for giving this a final push!

**Pull request merged**
